### DISCUSSION TITLE
Add convenience host functions for working with `String` and `Bytes`

### DISCUSSION THREAD

**Discussion by jayz22** [stellar/stellar-protocol#1511] [Core Advancement Proposals]
## What:

This discussion proposes adding several host functions for working with `String` and `Bytes`. 
I'll start by aggregating existing requests from https://github.com/stellar/rs-soroban-sdk/issues/1280, https://github.com/stellar/rs-soroban-env/issues/1187, https://github.com/stellar/rs-soroban-env/issues/1337, https://github.com/stellar/rs-soroban-env/issues/1101.
These are to be added under module "[buf](https://github.com/stellar/rs-soroban-env/blob/00ddd2714e757d0005bfc98798f05aa209f283bf/soroban-env-common/env.json#L1542)" (export = "b").

```
{
    "export": "n",
    "name": "string_append",
    "args": [
        {
            "name": "s1",
            "type": "StringObject"
        },
        {
            "name": "s2",
            "type": "StringObject"
        },
    ],
    "return": "StringObject",
    "docs": "Appends two strings `s1` and `s2` into a new string. Returns the combined string.",
    "min_supported_protocol": 22
},
{
    "export": "o",
    "name": "string_to_bytes",
    "args": [
        {
            "name": "s",
            "type": "StringObject"
        }
    ],
    "return": "BytesObject",
    "docs": "Returns a `BytesObject` containing the bytes from a string `s`.",
    "min_supported_protocol": 22
},
{
    "export": "p",
    "name": "string_from_bytes",
    "args": [
        {
            "name": "b",
            "type": "BytesObject"
        }
    ],
    "return": "StringObject",
    "docs": "Returns a `StringObject` containing the bytes from `b`.",
    "min_supported_protocol": 22
},
{
    "export": "q",
    "name": "bytes_first_index",
    "args": [
        {
            "name": "b",
            "type": "BytesObject"
        },
        {
            "name": "p",
            "type": "BytesObject"
        }
    ],
    "return": "Val",
    "docs": "Given a bytes `b`, and a search bytes `p`, return the first `U32Val` index of where the search bytes is found or `Void`. ",
    "min_supported_protocol": 22
},
{
    "export": "r",
    "name": "bytes_last_index",
    "args": [
        {
            "name": "b",
            "type": "BytesObject"
        },
        {
            "name": "p",
            "type": "BytesObject"
        }
    ],
    "return": "Val",
    "docs": "Given a bytes `b`, and a search bytes `p`, return the last `U32Val` index of where the search bytes is found or `Void`.",
    "min_supported_protocol": 22
},
{
    "export": "s",
    "name": "string_first_index",
    "args": [
        {
            "name": "s",
            "type": "StringObject"
        },
        {
            "name": "p",
            "type": "StringObject"
        }
    ],
    "return": "Val",
    "docs": "Given a string `s`, and a search string `p`, return the first `U32Val` index of where the search string is found or `Void`.",
    "min_supported_protocol": 22
},
{
    "export": "t",
    "name": "string_last_index",
    "args": [
        {
            "name": "s",
            "type": "StringObject"
        },
        {
            "name": "p",
            "type": "StringObject"
        }
    ],
    "return": "Val",
    "docs": "Given a string `s`, and a search string `p`, return the last `U32Val` index of where the search bytes is found or `Void`.",
    "min_supported_protocol": 22
},
```

## Why: 

These make working with `String` more convenient and efficient, without having to load bytes into the guest side.

**Comment by leighmcculloch** [comment]
All these functions are useful, the main ones of interest to me are the string<>bytes conversion host fns for efficient conversions between the two types that are essentially the same underlying type anyway. Today the only way to do that conversion is by copying to linear memory and either using an allocator to allocate that memory or buffering on some fixed memory.

Without these things existing it is almost preferrable to use Bytes over String wherever possible.

**Reply by jayz22** [reply]
Agreed. Alternatively we could just add the two string<>bytes conversion functions, and implement all algorithms (append, first, last) on bytes and convert to/from string whenever needed. This'll save on host function entries count, at the cost of additional calls for the conversation. We might do that in the future for uncommon ones (so we don't always have to provide two copies). But if these are common enough we can just add both versions.

**Comment by namankumar** [comment]
Thanks for aggregating the existing requests. That makes it simple to build a case for the proposal.

**Comment by tupui** [comment]
Oh yes that would be awesome 🙏
